syncer/types.go

```diff
@@ -0,0 +1,104 @@
+package syncer
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	rawGrpc "google.golang.org/grpc"
+
+	"github.com/0xPolygon/polygon-edge/blockchain"
+	"github.com/0xPolygon/polygon-edge/helper/progress"
+	"github.com/0xPolygon/polygon-edge/network"
+	"github.com/0xPolygon/polygon-edge/network/event"
+	"github.com/0xPolygon/polygon-edge/types"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"google.golang.org/protobuf/proto"
+)
+
+type Blockchain interface {
+	// SubscribeEvents subscribes new blockchain event
+	SubscribeEvents() blockchain.Subscription
+	// Header returns get latest header
+	Header() *types.Header
+	// GetBlockByNumber returns block by number
+	GetBlockByNumber(uint64, bool) (*types.Block, bool)
+	// VerifyFinalizedBlock verifies finalized block
+	VerifyFinalizedBlock(*types.Block) error
+	// WriteBlock writes a given block to chain
+	WriteBlock(*types.Block) error
+}
+
+type Network interface {
+	// AddrInfo returns Network Info
+	AddrInfo() *peer.AddrInfo
+	// RegisterProtocol registers gRPC service
+	RegisterProtocol(string, network.Protocol)
+	// Peers returns current connected peers
+	Peers() []*network.PeerConnInfo
+	// SubscribeCh returns a channel of peer event
+	SubscribeCh() (<-chan *event.PeerEvent, error)
+	// GetPeerDistance returns the distance between the node and given peer
+	GetPeerDistance(peer.ID) *big.Int
+	// NewProtoConnection opens up a new stream on the set protocol to the peer,
+	// and returns a reference to the connection
+	NewProtoConnection(protocol string, peerID peer.ID) (*rawGrpc.ClientConn, error)
+	// NewTopic Creates New Topic for gossip
+	NewTopic(protoID string, obj proto.Message) (*network.Topic, error)
+	// IsConnected returns the node is connecting to the peer associated with the given ID
+	IsConnected(peerID peer.ID) bool
+	// SaveProtocolStream saves stream
+	SaveProtocolStream(protocol string, stream *rawGrpc.ClientConn, peerID peer.ID)
+	// CloseProtocolStream closes stream
+	CloseProtocolStream(protocol string, peerID peer.ID) error
+}
+
+type Syncer interface {
+	// Start starts syncer processes
+	Start() error
+	// Close terminates syncer process
+	Close()
+	// GetSyncProgression returns sync progression
+	GetSyncProgression() *progress.Progression
+	// HasSyncPeer returns whether syncer has the peer syncer can sync with
+	HasSyncPeer() bool
+	// BulkSync syncs blocks to the peer's latest
+	BulkSync(context.Context, func(*types.Block) bool) error
+	// WatchSync starts routine to sync blocks
+	WatchSync(context.Context, func(*types.Block) bool) error
+}
+
+type Progression interface {
+	// StartProgression starts progression
+	StartProgression(startingBlock uint64, subscription blockchain.Subscription)
+	// UpdateHighestProgression updates highest block number
+	UpdateHighestProgression(highestBlock uint64)
+	// GetProgression returns Progression
+	GetProgression() *progress.Progression
+	// StopProgression finishes progression
+	StopProgression()
+}
+
+type SyncPeerService interface {
```

Comment:
Where's the other half of this interface (Close method) ?
Is it not needed at all ?

Reply:
I don't think module has to have Close method always. but fixed to close gRPC stream at 27eb6ab4dc3c3477bae54459f2b64e6240bc6a91